src/main/scala/workflow/NodeOptimizationRule.scala

```diff
@@ -5,96 +5,341 @@ import org.apache.spark.rdd.RDD
+        val pipeToSplice = oe.optimize(dataSample, numPerPartition).apply(dataSample)
+        val initialInstructionsToSplice = WorkflowUtils.pipelineToInstructions(pipeToSplice)
+        val dataSampleIndices = initialInstructionsToSplice.indices.filter {
+          initialInstructionsToSplice(_) == SourceNode(dataSample)
+        }.toSet
+        val instructionsToSplice = WorkflowUtils.disconnectAndRemoveInstructions(
+          dataSampleIndices.map(_ -> spliceEndpoint).toMap,
+          initialInstructionsToSplice
+        )._1
+
+        // First: Disconnect all the dependencies in the optimized pipeline on this existing EstimatorFitNode,
+        // and remove this node
+        val indexInNewPipeline = os.unoptimizedInstructionIndexToOptimizedIndex(index)
+        val removeResult = WorkflowUtils.disconnectAndRemoveInstructions(
+          Map(indexInNewPipeline -> (spliceEndpoint - 1)),
+          os.optimizedInstructions)
+        var partialOptimizedInstructions = removeResult._1
+        var partialOldIndexToNewIndex = os.unoptimizedInstructionIndexToOptimizedIndex andThen removeResult._2
+
+        // Then, insert the optimized fit & transform instructions to apply every place in the pipeline
+        // that was previously applying the output of this estimatorFitNode to new data
+        instructions.zipWithIndex.foreach {
+          case (TransformerApplyNode(t, tInputs), transformerApplyIndex) if t == index => {
+            // Disconnect all the dependencies in the optimized pipeline on this TransformerApplyNode,
+            // and replace them with the spliceEndpoint
+            val indexInNewPipeline = partialOldIndexToNewIndex(transformerApplyIndex)
+            val removeResult = WorkflowUtils.disconnectAndRemoveInstructions(
+              Map(indexInNewPipeline -> spliceEndpoint),
+              partialOptimizedInstructions)
+            partialOptimizedInstructions = removeResult._1
+            partialOldIndexToNewIndex = partialOldIndexToNewIndex andThen removeResult._2
+
+            // Splice in the instructions into the optimized pipeline
+            val spliceResult = WorkflowUtils.spliceInstructions(
+              instructionsToSplice,
+              partialOptimizedInstructions,
+              Map(spliceEndpoint -> partialOldIndexToNewIndex(inputIndices.head),
+                Pipeline.SOURCE -> partialOldIndexToNewIndex(tInputs.head)),
+              spliceEndpoint)
+            partialOptimizedInstructions = spliceResult._1
+            partialOldIndexToNewIndex = partialOldIndexToNewIndex andThen spliceResult._2
+          }
+
+          case _ => Unit
+        }
+
+        os.copy(
+          optimizedInstructions = partialOptimizedInstructions,
+          unoptimizedInstructionIndexToOptimizedIndex = partialOldIndexToNewIndex
+        )
+      }
+
+      case EstimatorOutput(oe: OptimizableLabelEstimator[a, b, l]) => {
+        val dataSpliceEndpoint = Pipeline.SOURCE - 1
```

Comment:
Will do! Numbers and order don't matter, the only things that matters are that they are different and don't refer to any existing instruction.